src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -922,30 +922,35 @@ MulticopterPositionControl::reset_alt_sp()
 void
 MulticopterPositionControl::limit_altitude()
 {
-	/* in altitude control, limit setpoint */
-	if (_run_alt_control && _pos_sp(2) <= -_vehicle_land_detected.alt_max) {
-		_pos_sp(2) = -_vehicle_land_detected.alt_max;
-		return;
-	}
+	if (_vehicle_land_detected.alt_max > 0.0f) {
 
-	/* in velocity control mode and want to fly upwards */
-	if (!_run_alt_control && _vel_sp(2) <= 0.0f) {
+		float altitude_above_home = -_pos(2) + _home_pos.z;
 
-		/* time to travel to reach zero velocity */
-		float delta_t = -_vel(2) / _acceleration_z_max_down.get();
+		// we are above maximum altitude
+		if (_run_alt_control &&  altitude_above_home > _vehicle_land_detected.alt_max) {
+			_pos_sp(2) = -_vehicle_land_detected.alt_max +  _home_pos.z;
+			return;
+		}
 
-		/* predicted position */
-		float pos_z_next = _pos(2) + _vel(2) * delta_t + 0.5f * _acceleration_z_max_down.get() * delta_t *delta_t;
+		// we want to fly upwards
+		if (!_run_alt_control && _vel_sp(2) <= 0.0f) {
 
-		if (pos_z_next <= -_vehicle_land_detected.alt_max) {
-			_pos_sp(2) = -_vehicle_land_detected.alt_max;
-			_run_alt_control = true;
-			return;
+			// time to reach zero velocity
+			float delta_t = -_vel(2) / _acceleration_z_max_down.get();
+
+			// predict position forward
```

Comment:
what does `position forward` mean? Please make a sentence describing this and explain the formula.